cookbooks/ros2_windows/recipes/qt5.rb

```diff
@@ -18,52 +21,68 @@
-end
+  # Update maintenance tool first, if necessary
+  windows_package 'Qt Maintenance Update' do
+    source 'c:\\Qt\\MaintenanceTool.exe'
+    installer_type :custom
+    # I couldn't find documentation, but return codes don't seem to correspond with an actual failure.
+    # Instead error information is written to the ErrorLogname below
+    returns [0]
+    options 'up qt.tools.maintenance'
+    timeout 600
+    only_if {::File.exist?('c:\\Qt\\MaintenanceTool.exe')}
+  end
 
-windows_package 'Qt Install' do
-  source 'http://download.qt.io/official_releases/online_installers/qt-unified-windows-x86-online.exe'
-  installer_type :custom
-  # I couldn't find documentation, but return codes don't seem to correspond with an actual failure.
-  # Instead error information is written to the ErrorLogname below
-  returns [0, 1, 3]
-  options '--script qt-installer.qs MsvcVersion=2019 ErrorLogname="' + error_filename + '"'
-  timeout 2000
-  not_if {::File.exist?('c:\\Qt\\MaintenanceTool.exe')}
+  windows_package 'Qt Maintenance' do
+    source 'c:\\Qt\\MaintenanceTool.exe'
+    installer_type :custom
+    # I couldn't find documentation, but return codes don't seem to correspond with an actual failure.
+    # Instead error information is written to the ErrorLogname below
+    returns [0, 1, 3]
+    options '--script qt-maintenance.qs MsvcVersion=2019 ErrorLogname="' + error_filename + '"'
+    timeout 2000
+    only_if {::File.exist?('c:\\Qt\\MaintenanceTool.exe')}
+  end
+
+  windows_package 'Qt Install' do
+    source 'http://download.qt.io/official_releases/online_installers/qt-unified-windows-x86-online.exe'
+    installer_type :custom
+    # I couldn't find documentation, but return codes don't seem to correspond with an actual failure.
+    # Instead error information is written to the ErrorLogname below
+    returns [0, 1, 3]
+    options '--script qt-installer.qs MsvcVersion=2019 ErrorLogname="' + error_filename + '"'
+    timeout 2000
+    not_if {::File.exist?('c:\\Qt\\MaintenanceTool.exe')}
+  end
+elsif node['ros2_windows']['qt5']['installation_method'] == 'offline'
+  qt_mirror_url = node['ros2_windows']['qt5']['qt_mirror_url']
+  error_filename = File.join(Dir.home(), "qt_install.err")
+  windows_package 'Qt Install' do
+    # Qt 5.12 was the last version of Qt to be provided an offline installer for open source usage.
+    source "#{qt_mirror_url}/official_releases/qt/5.12/5.12.10/qt-opensource-windows-x86-5.12.10.exe"
+    checksum "e315ec89000ae08d51c248f251ce670036b565e1d3a07df50e32557b8fd8246e"
+    installer_type :custom
+    # I couldn't find documentation, but return codes don't seem to correspond with an actual failure.
+    # Instead error information is written to the ErrorLogname below
+    returns [0, 1, 3]
+    options %{--st "#{qt_mirror_url}" --script qt-installer.qs --verbose MsvcVersion=2019 ErrorLogname="#{error_filename}"}
+    timeout 2000
+    not_if {::File.exist?('c:\\Qt\\Qt5.12.10\\MaintenanceTool.exe')}
```

Comment:
I think it's possible to specify the install directory of Qt, so they are the same for both offline and online.

Reply:
> it may now be possible to do a complete command-line unattended installer on windows

That's definitely a good item to put on the backlog: https://github.com/ros-infrastructure/ros2-cookbooks/issues/25